### PR TITLE
Enforce secure autoconfig and fetching from Exchange

### PR DESCRIPTION
**Pull request by boredsquirrel**
## Description
- enable autoconfig
- allow fetching from Exchange (needed for... exchange servers)
- do not empty the autoconfig and Addon URLs as this is irreversible
- add 2 hardening settings and enable them (even though "fetch from ISP" is still disabled)

## Reason and / or context
Autoconfig is very useful and not critical, as it contacts the mailserver only, to get infos. 

Contacting the ISP, or sending personal data is disabled, and the connection needs to be secure.

## How has this been tested ?
yes, on Thunderbird Flatpak.

## Types of changes :
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist :
- [x] My changes looks good ;
- [x] I agree that my code may be modified in the future ;
- [x] My code follows the code style of this project (see `.eslintrc.yml`).

### CODE REVIEW

File: user.js [hunk 1]
```diff
@@ -1153,14 +1153,17 @@ user_pref("_user.js.parrot", "9100 syntax error: this parrot is blind!");
  * if the user wishes to keep the existence of the mail provider private.
  * We also enforce (valid) SSL/TLS connections if auto-configuration happens to be enabled.
  * [1] https://developer.mozilla.org/en-US/docs/Mozilla/Thunderbird/Autoconfiguration ***/
-user_pref("mailnews.auto_config.guess.enabled", false);
+// user_pref("mailnews.auto_config.guess.enabled", false);
```
Comment (HorlogeSkynet) [inline user.js:1156]: I'm not sure about keeping autoconfig enabled, **by default**.
This project is a template, aiming to provide an hardened version of Thunderbird, and autoconfig stays one of these features explicitly _leaking_ future usages information to third-parties (even if they don't have malicious intents, their servers receive such data) like Mozilla or ISP.
If we want TB to "simply" speaks IMAP/POP/SMTP, we can't expect it to send mail server addresses _away_ beforehand.

When we use a giant email providers (like Google or Hotmail), it's not necessarily an issue (from a privacy point of view here), but if your email server is somehow specific and maybe not public (self-hosted, proposed by company, or maybe a Tor hidden service ?), I guess this becomes problematic.

File: user.js [hunk 2]
```diff
@@ -1153,14 +1153,17 @@ user_pref("_user.js.parrot", "9100 syntax error: this parrot is blind!");
...
 user_pref("mailnews.auto_config.fetchFromISP.enabled", false);
 user_pref("mailnews.auto_config.fetchFromISP.sendEmailAddress", false);
-user_pref("mailnews.auto_config.fetchFromExchange.enabled", false);
+// user_pref("mailnews.auto_config.fetchFromExchange.enabled", false);
```
Comment (HorlogeSkynet) [inline user.js:1159]:
```suggestion
   // user_pref("mailnews.auto_config.fetchFromExchange.enabled", false);
```

(disabled preferences are left-padded a bit)

File: user.js [hunk 3]
```diff
@@ -1153,14 +1153,17 @@ user_pref("_user.js.parrot", "9100 syntax error: this parrot is blind!");
...
 user_pref("mailnews.auto_config.guess.sslOnly", true);
 user_pref("mailnews.auto_config.guess.requireGoodCert", true); // [DEFAULT: true]
-user_pref("mailnews.auto_config_url", "");
-user_pref("mailnews.auto_config.addons_url","");
+// user_pref("mailnews.auto_config_url", "");
+// user_pref("mailnews.auto_config.addons_url","");
```
Comment (HorlogeSkynet) [inline user.js:1163]:
```suggestion
   // user_pref("mailnews.auto_config_url", "https://live.thunderbird.net/autoconfig/v1.1/");
   // user_pref("mailnews.auto_config.addons_url","https://live.thunderbird.net/autoconfig/addons.json");
```

(same here, but let's include default values)

File: user.js [hunk 4]
```diff
@@ -1153,14 +1153,17 @@ user_pref("_user.js.parrot", "9100 syntax error: this parrot is blind!");
...
+user_pref("mailnews.auto_config.fetchFromISP.sslOnly", true);
+user_pref("mailnews.auto_config.fetchFromExchange.sslOnly", true);
```
Comment (HorlogeSkynet) [inline user.js:1165]:
```suggestion
user_pref("mailnews.auto_config.guess.sslOnly", false);
user_pref("mailnews.auto_config.guess.requireGoodCert", true);
```

Let's move ISP/SSL "up", and also enforce these preferences too.
Although, `fetchFromExchange.sslOnly` does not seem to exist (at least against TB v115).

File: user.js [hunk 5]
```diff
@@ -1153,14 +1153,17 @@ user_pref("_user.js.parrot", "9100 syntax error: this parrot is blind!");
  * if the user wishes to keep the existence of the mail provider private.
  * We also enforce (valid) SSL/TLS connections if auto-configuration happens to be enabled.
  * [1] https://developer.mozilla.org/en-US/docs/Mozilla/Thunderbird/Autoconfiguration ***/
-user_pref("mailnews.auto_config.guess.enabled", false);
+// user_pref("mailnews.auto_config.guess.enabled", false);
 user_pref("mailnews.auto_config.fetchFromISP.enabled", false);
```
Comment (HorlogeSkynet) [inline user.js:1157]:
```suggestion
user_pref("mailnews.auto_config.fetchFromISP.enabled", false);
user_pref("mailnews.auto_config.fetchFromISP.sslOnly", true);
```

**Review by HorlogeSkynet — CHANGES_REQUESTED** [review]
Please find my comments inline. Thanks for your time :pray:

File: user.js (hunk 1 above)
Comment (boredsquirrel) [reply user.js:1156]: yes for sure. My idea was to add commented out lines for more security and privacy if the ISP would be used (for whatever reason).

But to my knowledge autoconfig is used to get the config from the mail provider, so the request is sent to their servers, not some random others.

If this is not true, many changes can be reverted, as contacting random 3rd parties makes little sense

**Comment by boredsquirrel** [comment]
Hi, sorry for the late reply. Added most changes that are clear to me.

so, deactivated but useful settings get no indentation, deactivated and obsolete settings get an indentation?

**Comment by HorlogeSkynet** [comment]
> Hi, sorry for the late reply.

Hey, sorry too. I'll try to review all subjects in a batch again.

> so, deactivated but useful settings get no indentation, deactivated and obsolete settings get an indentation?

Disabled preferences are always indented. Enabled ones aren't.
Deprecated (upstream) settings follow the same rule.

File: user.js [hunk 6]
```diff
@@ -1153,14 +1153,16 @@ user_pref("_user.js.parrot", "9100 syntax error: this parrot is blind!");
...
-user_pref("mailnews.auto_config.fetchFromExchange.enabled", false);
+user_pref("mailnews.auto_config.fetchFromISP.sslOnly", true);
+    // user_pref("mailnews.auto_config.fetchFromExchange.enabled", false);
 user_pref("mailnews.auto_config.guess.sslOnly", true);
 user_pref("mailnews.auto_config.guess.requireGoodCert", true); // [DEFAULT: true]
-user_pref("mailnews.auto_config_url", "");
-user_pref("mailnews.auto_config.addons_url","");
+    // user_pref("mailnews.auto_config_url", "https://live.thunderbird.net/autoconfig/v1.1/");
+    // user_pref("mailnews.auto_config.addons_url","https://live.thunderbird.net/autoconfig/addons.json");
+
```
Comment (HorlogeSkynet) [inline user.js:1165]:
```suggestion
```

(very picky : a newline got added here)

File: user.js (hunk 1 above)
Comment (HorlogeSkynet) [reply user.js:1156]: > But to my knowledge autoconfig is used to get the config from the mail provider, so the request is sent to their servers, not some random others.

Unfortunately it doesn't work that way : a query is sent to Mozilla servers to gather information about mail providers own servers (cf. `mailnews.auto_config_url` value).

> If this is not true, many changes can be reverted, as contacting random 3rd parties makes little sense

I am okay with other changes, as long as we keep AC disabled by default :wink: 

PS : I've noticed there is an explicit AC mention in script top comment, pointing to `9101` and `9102` re-enabling if needed.

Comment (HorlogeSkynet) [reply user.js:1156]: So I've reverted this specific part according to above point.

**Review by HorlogeSkynet — APPROVED** [review]

**Pull request merged**